### PR TITLE
Ruby bindings for Rust foundation (magnus)

### PR DESCRIPTION
**Pull request by jychen7**
close https://github.com/datafusion-contrib/datafusion-ruby/issues/1

This PR is a foundation of Ruby bindings for Rust
- It creates the `Datafusion::SessionContext` class and its initializer
- README is added for the roadmap
- CONTRIBUTING is added for developer guide and FAQ about choices
- Github Action is added, see a successful run here at https://github.com/jychen7/datafusion-ruby/actions/runs/2604878357

### CODE REVIEW

**Comment by andygrove** [comment]
Hi @kou could you take a look? Thanks.

@jychen7 There is a tracking issue for Ruby bindings - https://github.com/apache/arrow-datafusion/issues/1114

**Comment by jychen7** [comment]
@kou sorry, I didn't realize there is discussion at https://github.com/apache/arrow-datafusion/issues/1114 and also https://github.com/apache/arrow-datafusion/issues/1113 about using `Rust -> C -> Ruby/Java/etc`

just fyi, for `Rust -> Ruby`, I have the basic `register_csv` and `sql` PoC at https://github.com/jychen7/datafusion-ruby/tree/dev. Happy to discuss which path we want to go with.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+# FAQ
+
+## Why Magnus?
+
+As of 2022-07, there are a few popular Ruby bindings for Rust, [Rutie](https://github.com/danielpclark/rutie), [Magnus](https://github.com/matsadler/magnus) and [other alternatives](https://github.com/matsadler/magnus#alternatives). Magnus is picked because its API seems cleaner and it seems more clear about safe vs unsafe. The author of Magnus have a "maybe bias" comparison in this [reddit thread](https://www.reddit.com/r/ruby/comments/uskibb/comment/i98rds4/?utm_source=share&utm_medium=web2x&context=3). It is totally subjective and it should not be large effort if we decides to switch to different Ruby bindings fr Rust in future.
+
+## Why the module name and gem name are different?
+
+The module name `Datafusion` follows the [datafusion](https://github.com/apache/arrow-datafusion) and [datafusion-python](https://github.com/datafusion-contrib/datafusion-python). The gem name `datafusion` [is occupied in rubygems.org at 2016](https://rubygems.org/gems/datafusion), so our gem is called `arrow-datafusion`.
```
Comment (kou) [inline CONTRIBUTING.md:40]: I have a plan to use `DataFusion` for module name and `red-datafusion` for gem name.

**Comment by kou** [comment]
@jychen7 Thanks for working on this!

Yes. I want to use the `Rust -> C -> Ruby/Java/etc` approach. I think that this approach is better for DataFusion community because DataFusion community can work together on C layer and share the work with multiple languages including Ruby. It will reduce development resource for the Ruby bindings. (It will share development resource with Ruby and other languages.) Unfortunately, there aren't many people who want to develop data processing related gems in Ruby community at present. So I want to work together with people not in Ruby community too. What do you think about this?

See also my talk at [The Data Thread](https://thedatathread.com/):
* Video: https://www.youtube.com/watch?v=ex6qoj8agvk
* Slide: https://slide.rabbit-shocker.org/authors/kou/the-data-thread/

We have an implementation for `DataFusion::SessionContext#sql` in the `Rust -> C -> Ruby/Java/etc` approach: https://github.com/datafusion-contrib/datafusion-c/blob/main/test/test-data-frame.rb

And I'm working on adding support for Apache Arrow C data interface: https://github.com/datafusion-contrib/datafusion-c/issues/9

It almost works on my local environment. I'll complete it soon. Please wait for a while. With the work, we can integrate DataFusion with Red Arrow like the following:

```ruby
context = DataFusion::SessionContext.new
record_batch = Arrow::RecordBatch.new(booleans: [true, false, nil],
                                      integers: [1, nil, 3])
_success, c_abi_array, c_abi_schema = record_batch.export
context.register_record_batch("data", c_abi_schema, c_abi_array)
data_frame = context.sql("SELECT * FROM data")
p(record_batch.to_table == data_frame.to_table) #=> true
```

I'll also open a pull request for the `Rust -> C -> Ruby/Java/etc` approach in this repository. Please wait for a while.

**Comment by jychen7** [comment]
> DataFusion community can work together on C layer and share the work with multiple languages including Ruby

It makes sense. I understand the advantages of "unify" the core for different language bindings.

> Unfortunately, there aren't many people who want to develop data processing related gems in Ruby community at present. So I want to work together with people not in Ruby community too

yes, python and java are quite popular in both batch world (Spark) and streaming world (e.g. Flink).

> Why Apache Arrow is important for Ruby

Thanks for the great sharing. ADBC and Arrow Flight SQL can definitely benefit ruby web app. However, I am curious have you see any potential use cases of DuckDB or Datafusion in ruby applications, either web app or background worker?

**Comment by jychen7** [comment]
btw, I am going to close this PR then. What do you think?
I will keep the `Rust -> Ruby` implementation  in personal repo as side project. It is just my trial to play with Magnus.

File: CONTRIBUTING.md (hunk 1 above)
Comment (jychen7) [reply CONTRIBUTING.md:40]: > red-datafusion

good name, another red-data-tools! I think of this name as well, of course, inspire by your `red-arrow`.

btw, during this PR, I have publish as `arrow-datafusion` to rubygems.org. Let me know if `datafusion-contrib` want to re-claim the ownership

**Comment by kou** [comment]
> I am curious have you see any potential use cases of DuckDB or Datafusion in ruby applications, either web app or background worker?

I want to use them as data frame API backends like Ibis+DuckDB in Python and dplyr+DuckDB in R as the first step. I think that Ruby community needs at least one Apache Arrow based data frame library. If Ruby community has an Apache Arrow based data frame library, we can mix Ruby and other languages to process data. We can use an Apache Arrow based data frame library to interchange data from/to Ruby. For example, Apache Spark and PySpark interchanges data frame as Apache Arrow data.

This is just an idea. We may have better approach than this to use Ruby for data processing...

FYI: Red Data Tools is working on ActiveRecord+DuckDB as another approach:  https://github.com/red-data-tools/activerecord-duckdb-adapter

> btw, I am going to close this PR then. What do you think?
> I will keep the `Rust -> Ruby` implementation in personal repo as side project. It is just my trial to play with Magnus.

OK. How about adding a link for your repository as another approach to the `README.md` of this repository?

File: CONTRIBUTING.md (hunk 1 above)
Comment (kou) [reply CONTRIBUTING.md:40]: Thanks!
If you don't mind, could you add me to the owners of `arrow-datafusion` for future transfer (that may not be occurred...)?

```bash
$ gem owner -a kou arrow-datafusion
```

If you want to release more versions from https://github.com/jychen7/datafusion-ruby in the future, you don't need to add me to the owners of `arrow-datafusion`.

**Comment by jychen7** [comment]
> How about adding a link for your repository as another approach to the README.md of this repository?

sounds good, I will create a PR to link it after your https://github.com/datafusion-contrib/datafusion-ruby/pull/4 is merged.

---

I have separate my alternative implementation to https://github.com/jychen7/arrow-datafusion-ruby and do you think the FAQ "Why another Ruby bindings for Arrow Datafusion?" in https://github.com/jychen7/arrow-datafusion-ruby/pull/1 make senses to you? Feel to add comment there

File: CONTRIBUTING.md (hunk 1 above)
Comment (kou) [reply CONTRIBUTING.md:40]: OK.

**Comment by kou** [comment]
> sounds good, I will create a PR to link it after your https://github.com/datafusion-contrib/datafusion-ruby/pull/4 is merged.

Thanks. I merged #4. Could you open a pull request for it?

> I have separate my alternative implementation to https://github.com/jychen7/arrow-datafusion-ruby and do you think the FAQ "Why another Ruby bindings for Arrow Datafusion?" in https://github.com/jychen7/arrow-datafusion-ruby/pull/1 make senses to you? Feel to add comment there

OK. I'll left a comment to https://github.com/jychen7/arrow-datafusion-ruby/pull/1 .

**Pull request closed without merging**